File: keyple-example/calypso/android/omapi/src/main/java/org/eclipse/keyple/example/android/omapi/MainActivity.java
/********************************************************************************
 * Copyright (c) 2018 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * See the NOTICE file(s) distributed with this work for additional information regarding copyright
 * ownership.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
package org.eclipse.keyple.example.android.omapi;

import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;


/**
 * Example of Android application to test
 * 
 * @{@link org.eclipse.keyple.plugin.android.omapi.AndroidOmapiPlugin}
 *
 */
public class MainActivity extends AppCompatActivity {


    private static final String TAG = MainActivity.class.getSimpleName();
    private static final String TAG_OMAPI_TEST_VIEW = "tagomapitestfragment";


    private DrawerLayout mDrawerLayout;

    /**
     * Setup navigation drawer
     * 
     * @param savedInstanceState
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        /**
         * Define UI elements
         */
        Log.d(TAG, "onCreate");
        setContentView(R.layout.activity_main);

        setupNavDrawer();
        /**
         * end of Define UI elements
         */
        activateOMAPITestView();

    }

    /**
     * Inject OMAPI Test View fragment into activity
     */
    private void activateOMAPITestView() {
        // init OMAPI Test Fragment
        Log.d(TAG, "Insert OMAPI Test View  Fragment");
        OMAPITestFragment omapiTestFragment = OMAPITestFragment.newInstance();
        getFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, omapiTestFragment, TAG_OMAPI_TEST_VIEW)
                .addToBackStack(null).commit();
    }

    /**
     * Configure Navigation Drawer
     */
    private void setupNavDrawer() {
        mDrawerLayout = findViewById(R.id.drawer_layout);

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(MenuItem menuItem) {
                        // set item as selected to persist highlight
                        menuItem.setChecked(true);
                        // close drawer when item is tapped
                        mDrawerLayout.closeDrawers();

                        // Add code here to update the UI based on the item selected
                        // For example, swap UI fragments here
                        Log.d(TAG, "Item selected from drawer: " + menuItem.getTitle());

                        switch (menuItem.getItemId()) {
                            case R.id.nav_omapi:
                                activateOMAPITestView();
                                break;
                            default:
                                throw new IllegalArgumentException("Invalid menuItem");

                        }
                        return true;
                    }
                });

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionbar = getSupportActionBar();
        actionbar.setDisplayHomeAsUpEnabled(true);
        actionbar.setHomeAsUpIndicator(R.drawable.ic_menu_black_24dp);



    }

    // 'Open' event for navigation drawer
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: java/example/calypso/android/omapi/src/main/kotlin/org/eclipse/keyple/example/calypso/android/omapi/activity/MainActivity.kt
/********************************************************************************
 * Copyright (c) 2020 Calypso Networks Association https://www.calypsonet-asso.org/
 *
 * See the NOTICE file(s) distributed with this work for additional information regarding copyright
 * ownership.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse
 * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
package org.eclipse.keyple.example.calypso.android.omapi.activity

import android.content.Intent
import android.os.Bundle
import android.view.View
import kotlinx.android.synthetic.main.activity_main.calypsoExampleButton
import kotlinx.android.synthetic.main.activity_main.connectOmapiTV
import kotlinx.android.synthetic.main.activity_main.coreExamplesButton
import kotlinx.android.synthetic.main.activity_main.menuLayout
import kotlinx.android.synthetic.main.activity_main.progressBar
import kotlinx.android.synthetic.main.activity_main.toolbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.eclipse.keyple.core.seproxy.Reader
import org.eclipse.keyple.core.seproxy.SmartCardService
import org.eclipse.keyple.core.seproxy.exception.KeyplePluginInstantiationException
import org.eclipse.keyple.core.seproxy.exception.KeyplePluginNotFoundException
import org.eclipse.keyple.example.calypso.android.omapi.R
import org.eclipse.keyple.plugin.android.omapi.PLUGIN_NAME
import timber.log.Timber

class MainActivity : BasicActivity(), View.OnClickListener {

    companion object {
        private const val MAX_TRIES = 10
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initActionBar(toolbar, "OMAPI Plugin", "Examples application")

        coreExamplesButton.setOnClickListener(this)
        calypsoExampleButton.setOnClickListener(this)
    }

    override fun onResume() {
        super.onResume()
        // Retrieve OMAPI Readers, can't observed but may take time.
        // So we retry every second 10 times
        progressBar.visibility = View.VISIBLE

        CoroutineScope(Dispatchers.Main).launch {
            try {
                connectOmapi()
                // readers are initialized, we can show the menu
                connectOmapiTV.visibility = View.GONE
                menuLayout.visibility = View.VISIBLE
            } catch (e: KeyplePluginNotFoundException) {
                Timber.e(e)
                showAlertDialog(e)
            } catch (e: KeyplePluginInstantiationException) {
                Timber.e(e)
                showAlertDialog(e)
            } finally {
                progressBar.visibility = View.INVISIBLE
            }
        }
    }

    @Throws(KeyplePluginNotFoundException::class, KeyplePluginInstantiationException::class)
    private suspend fun connectOmapi(): Map<String, Reader> = withContext(Dispatchers.IO) {
        var readers: Map<String, Reader> ? = null
        for (x in 1..MAX_TRIES) {
            readers = SmartCardService.getInstance().getPlugin(PLUGIN_NAME).readers
            if (readers == null || readers.size < 1) {
                Timber.d("No readers found in OMAPI Keyple Plugin")
                Timber.d("Retrying in 1 second")
                delay(1000)
            } else {
                Timber.d("Readers Found")
                break
            }
        }

        readers ?: throw KeyplePluginInstantiationException(getString(R.string.error_no_reader_found, MAX_TRIES))
    }

    override fun onClick(v: View?) {
        v?.let {
            when (it.id) {
                R.id.calypsoExampleButton -> startActivity(Intent(this, CalypsoExamplesActivity::class.java))
                R.id.coreExamplesButton -> startActivity(Intent(this, CoreExamplesActivity::class.java))
                else -> {}
            }
        }
    }
}
